src/inference/AdvancedSMC.jl

```diff
@@ -383,34 +383,32 @@ function resample_multinomial(w::AbstractVector{<:Real}, num_particles::Integer)
 end
 
 function resample_residual(w::AbstractVector{<:Real}, num_particles::Integer)
-
-    M = length(w)
-
+    m = length(w)
+    @assert m==num_particles
```

Comment:
This doesn't have to be the case (and that's the reason why some of the code below has to be changed)
```suggestion
```
BTW I believe this was an additional bug in the previous implementation since it used `M = length(w)` where `num_particles` would have been more appropriate.